go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1010,30 +1010,64 @@ func (tsv *TabletServer) Execute(ctx context.Context, target *querypb.Target, sq
 			if err != nil {
 				return err
 			}
-			qre := &QueryExecutor{
-				query:          query,
-				marginComments: comments,
-				bindVars:       bindVariables,
-				transactionID:  transactionID,
-				options:        options,
-				plan:           plan,
-				ctx:            ctx,
-				logStats:       logStats,
-				tsv:            tsv,
-			}
-			extras := tsv.watcher.ComputeExtras(options)
-			result, err = qre.Execute()
-			if err != nil {
-				return err
+			if plan.Reason == planbuilder.ReasonTopic {
+				result, err = tsv.topicExecute(ctx, query, comments, bindVariables, transactionID, options, plan, logStats)
+			} else {
+				result, err = tsv.qreExecute(ctx, query, comments, bindVariables, transactionID, options, plan, logStats)
 			}
-			result.Extras = extras
-			result = result.StripMetadata(sqltypes.IncludeFieldsOrDefault(options))
-			return nil
+
+			return err
 		},
 	)
 	return result, err
 }
 
+func (tsv *TabletServer) topicExecute(ctx context.Context, query string, comments sqlparser.MarginComments, bindVariables map[string]*querypb.BindVariable, transactionID int64, options *querypb.ExecuteOptions, plan *TabletPlan, logStats *tabletenv.LogStats) (result *sqltypes.Result, err error) {
+	if plan.PlanID != planbuilder.PlanInsertTopic {
+		return nil, vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "Only inserts allowed on topics")
+	}
+
+	// choose InsertMessage as the new PlanID
+	plan.PlanID = planbuilder.PlanInsertMessage
```

Comment:
plan is accessed concurrently by multiple goroutines. So, it should be treated as read-only in this section of the code. I also don't see a reason to change the plan here. It looks like you're building individual `InsertMessage` plans for each subsriber below anyways.